Improve Parameter Syntax

mitchphillipson opened the pull request:
Fixes Issue #155

mitchphillipson commented:
I refactored this slightly, added some tests, and improved the docstrings. 

@EliLazarus I believe this is ready for review.

mitchphillipson commented:
The update parameter syntax and features were missing from the other variables (like sectors). I updated the other MPSGEvariables to match the parameter.

I also updated how models are printed. For example, for the given sector:

```julia
@sector(M, S[regions = R], description = "Demonstration Sector")
```
The model would print the following:
```
$Sectors:
S[:r1]
S[:r2]
S[:r3]
```
This is fine for small models, but terrible for large models. Now it will print
```
$Sectors:
S[regions] -- Demonstration Sector
```

Also made the `parameters` and `auxiliary` variables only print if they exist. So you won't have an empty `$Parameters` block.

EliLazarus commented on src/macros/@parameter.jl:
```diff
@@ -121,6 +167,33 @@ macro parameter(input_args...)
+    Keyword arguments must be contained within parentheses (refer to the example
+    above).
+"""
 macro parameters(model, block)
```
These are great!

EliLazarus commented:
Mid-review questions and comments.
Works locally and passes all tests for me. Seems good.
It's a breaking change with the new index syntax, so required updating other scripts/models (fine).
It's a breaking change re the variable name to Symbol. Why that change?
And why paramter name => string in structs?

I would benefit from either 1) more explanation comments on the new code, or 2) more comments on the changes in the repo/PR, or 3) going over parts of the PR as a discussion (easist?).
I would also benefit from another validation/explanation of the change (esp bc it's breaking). Maybe a comment here in the PR.

mitchphillipson commented:
1. The indexing change is moderately breaking. The `index` keyword has been removed in favor of the new syntax. Any code not using the `index` keyword still operates as expected.
2. The reason for the `Symbol` => `String` change is there is a JuMP function `Containers.build_name_expr` that builds a string by default. I figured it was better to defer to outputs directly from JuMP. The function getters, however, still return symbols. This is the main reason it's advised to not use field names directly. Up to discussion on how to proceed.

I think it may be better to put the discussion of the changes directly in the associated issue #155. What are you looking for to be detailed?

mitchphillipson commented:
1. Added tests to ensure the correct errors are thrown for a given variable.
2. Documented how to set start values, lower bounds, and upper bounds on auxiliary variables
3. Moved `add_variable` to the `add!`, which means a variable can't be added to the model without a corresponding JuMP variable
4. Made `string_name` return a string
5. Added constructors to variables for optional descriptions
6. Documented how the indices get printed in variables

EliLazarus commented on src/macros/@MPSGEvariable.jl:
```diff
@@ -466,6 +498,11 @@ M = MPSGEModel()
+
+!!! note
+    By default auxiliary variables start at 0 with on lower or upper bounds. These
+    can be set after variable creating using [`set_start_value`](@ref), 
```
I'm confused by "0 with on lower or upper bounds".
My undertanding: By default, there's a 0 lower bound. And 0 is (also) the default start value. 
I don't see how the upper bounds is relevant here. And is "on" a typo?

mitchphillipson replied:
"on" should be "no". 

You can see this in the `add_variable!(m::MPSGEModel, S::Auxiliary)` function in `build.jl`. Only the start value of 0 is set. This matches the behavior of MPSGE in GAMS, any bounds on auxiliary variables are set after creation.

This is only true for auxiliary variables, all other (non-parameter) variables get a lower bound of 0.

EliLazarus commented:
Ok, excellent. I think it's good to go, as far as I can tell.
Tag and merge I reckon.

Also noting another breaking change related to the strings. string.(MPSGEVariable) now prints as "Var[index]", where before it was "Y(index,)". The update is better, but it is a breaking change for me because I happen to parse and slice that string in a script.

I think we lost the add!(Model, Variabletype) API function version here, and I think that should be an option. My understanding is that the add! now just adds everything to JuMP, and the macros build the local model. 

If that's true, and it's easy to add back, before merge is great. If not easy, merge, add as an issue, and get to it later. And if I'm wrong, I stand corrected and apologise.
